(#382) Definition of Done, User Story, and Acceptance Criteria (#383)
chocolatey/home: definition-of-done.md

# Definition of Done

The **Definition of Done** is all about the big picture, focusing on the overall quality and completeness of the project, feature, or [User Story](./user-story-and-acceptance-criteria.md#user-story). It’s a shared understanding among the team that sets the standard for what "done" means. The Definition of Done includes criteria like code quality, testing, documentation, and integration.

## The Outcome is Achieved

- [ ] The [Acceptance Criteria](./user-story-and-acceptance-criteria.md#acceptance-criteria) for the item are fully met.

> - [ ] The work delivers the intended user or business value.

Reviewer: ```suggestion
- [ ] The work delivers user, or business, value.
```

Reviewer: This may seem pedantic, but I had to re-read what was originally written several times to grasp what was meant. This is just my suggestion on how ot make it clearer.

Author: Hmm.. to make more clear I might suggest we change it to `The work delivers the user or business value defined by the User Story.` (and link to the definition of a User Story). It's not so much of "does it deliver value", but more of "does it deliver the value that was defined".  There's a chance that we could be delivering value, but we missed the mark on what the issue or feature was really requesting.

Author: This change has been made.

- [ ] Edge cases, error handling, and failure states are handled or explicitly documented.

## The Code Is Production-Ready

- [ ] Code follows team standards and linting rules.
- [ ] No new warnings, lint errors, or build failures are introduced.

> - [ ] No leftover debug code, TODOs, or commented-out blocks without clear notes.

Reviewer: ```suggestion
- [ ] No leftover debug code, TODOs, or commented-out blocks without including the date, who made the change and a clear why.
```

Author: This change has been made.

> - [ ] Dependencies and licenses are correctly recorded (CREDITS.json if applicable).

Author: I have changed this a bit, to `License information for product dependencies are correctly recorded in the `CREDITS.md`, `CREDITS.pdf`, and `CREDITS.json` files.` this is because not all repositories will have these files, but products will, and in those these files must be correct.

## It Is Tested and Verified

- [ ] Automated tests (unit, integration, Playwright, etc.) are added or updated as needed.
- [ ] All CI/CD checks pass.

> - [ ] The change has been verified in a real environment (locally, preview, or staging).

Reviewer: Is this referring to the submission of a PR / MR? Locally verifying a change is not sufficient to have it merged, but is sufficient to submit a PR / MR.

Author: > Locally verifying a change is not sufficient to have it merged

I would agree with this for products definitely, but for our websites like docs, blog, chocolatey.org, design, and a few others, locally viewing the website is what we do before merging.

Maybe this point could be changed to something like `The change has been verified in a staging environment for Chocolatey products and the Chocolatey Community Repository,  and locally for additional websites.`

Reviewer: > but for our websites like docs, blog, chocolatey.org, design, and a few others, locally viewing the website is what we do before merging.

I did not think of that. Good point.

> Maybe this point could be changed to something like The change has been verified in a staging environment for Chocolatey products and the Chocolatey Community Repository,  and locally for additional websites.

Is being more specific better, or adds complexity?

Author: I think there's a fine line here of being too specific or not a specific enough. I'll give this wording some thought. I do think it needs to change somehow though so we can easily determine what environment we expect it to be verified in.

Reviewer 2: What about `The change has been verified in the product's defined verification environment.`? This would require each repo to specify it's verification environment, but it would at least allow this document to be generic and each repo to be specific.

Alternatively something about static versus dynamic websites?

Author: I went with @corbob's suggestion here. We can always adjust this as we see fit in the future.

## It Is Accessible and Usable

- [ ] Keyboard navigation works.
- [ ] Semantic markup is used appropriately.
- [ ] No new accessibility violations are introduced.

## It Is Reviewed

> - [ ] At least one peer has reviewed and approved the work.

Reviewer: ```suggestion
- [ ] At least one human peer has reviewed and approved the work.
```

Author: This change has been made.

> - [ ] Code Owner approval exists when dependencies are removed, added, or updated.

Reviewer: ```suggestion
- [ ] Code Owner approval is required when dependencies are removed, added, or updated.
```

Author: Could I suggest something more along the lines below? This would make it more inline with how the first point is worded. Moving it underneath (inside?) the top point also prevents any top level checkboxes from remaining unchecked (in the case of no dependency changes).

```
- [ ] At least one peer has reviewed and approved the work.
    - [ ] At least one Code Owner has reviewed and approved any dependencies that have been removed, added, or updated.
  ```

Author: This change has been made.

- [ ] All review comments are addressed or explicitly resolved.

## It Is Understandable

> - [ ] Anyone can trace why the change was made and what it delivered.

Reviewer: Is 'trace' the right word here? I didn't make a suggestion, as I may be misunderstanding the context.

If it's not, would 'determine' or 'understand' be better?

Author: The meaning of this one is that a person can see a change and then trace it back to the PR that introduced it, and the commit, and all the way back to the issue that defined why changed in the first place. More like a "paper trail". I think the word `determine`  would fit well.

Author: This change has been made.

> - [ ] Any necessary documentation, screenshots, or notes have been updated.

Reviewer: ```suggestion
- [ ] Any necessary documentation, screenshots, or notes have been added or updated.
```

Author: This change has been made.

> - [ ] Future maintainers can reason about the change without tribal knowledge.

Reviewer: ```suggestion
- [ ] Future code maintainers can reason about the change without tribal knowledge.
```

Reviewer: I added this because we often talk about 'maintainers' in the context of 'package maintainers' and not 'code maintainers'.

Author: This change has been made.
